vision/imageloader/imageloader.go

```diff
@@ -259,3 +264,22 @@ func decodeImage(buffer []byte, colorSpace string) (gocv.Mat, error) {
 	}
 	return m, e
 }
+
+func init() {
+	// We use these background threads to call `gocv`. This is because `gocv` makes
+	// `Cgo` calls extensively, if we call `gocv` directly in goroutines(each epoch
+	// creates a new goroutine), the `Cgo` calls will cause Go runtime to create too
+	// many threads.
+	go func() {
+		runtime.LockOSThread()
+		for f := range readSamplesCh {
+			f()
+		}
+	}()
+	go func() {
```

Comment:
What will happen if we create two ImageLoader instance?

Reply:
Because there's only one OS thread now, the second `ImageLoader` will start to load the file after the first one is finished.

Reply:
Could we directly call `readSamples` function?

```go
go func() {
    runtime.LockOSThread()
    m.readSamples()
}

go func() {
    runtime.LockOSThread()
    m.samplesToMinibatches()   
}
```

Reply:
This will cause a memory leak in the current version of OpenCV, see https://github.com/opencv/opencv/issues/9745 and https://github.com/opencv/opencv/pull/15692#issuecomment-619334663

Reply:
@shendiaomo It seems that there is much magic in this code change. We may need to add more comments. Why we use channels here? Why we use lock here? The comments would be very helpful.

Reply:
Done.